proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,222 @@
+This service starts automatically when a validator starts and listens to a
+designated channel for any incoming rewards calculation requests, performing
+the necessary rewards calculations accordingly.
+
+When the block on the network crosses the epoch boundary, a reward computation
+request based on stake delegation and vote states of vote accounts at the epoch
+boundary is sent to the `EpochRewardCalculationService` to start the process of
+epoch reward computation. This marks the start of the reward computation phase.
+
+The reward computation phase lasts for `N` blocks since the beginning of the
+epoch boundary. When reaching block height `N` after the start of the
+`reward calculation phase`, the bank will mark it the end the `reward calculation phase`.
+A conservative value of 1,000 is chosen for `N` to accommodate the most of the
+validators on the network.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will block and wait until the computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase (end of block height `N`), a set of
+sysvar accounts will be updated - `EpochRewardProof` and `RewardReserve`.
+
+`EpochRewardProof` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
+```
+, where each entry corresponds to the reward for
+one particular epoch. The maximum depth of the history are the same as the
+depth of stake history, which is 512.
+
+When reaching the end of reward computation phase, `EpochRewardProof` are
+updated with the reward computation result. A new entry will be pushed into the
+array. If the length of the history array reaches the maximum depth, the oldest
+entry will popped.
+
+```
+struct RewardHistoryEntry {
+   total: u64,                  // total rewards for this epoch
+   distributed: u64,            // already distributed reward amount
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+At the end of reward computation phase, a set of `EpochRewardsReserve` sysvar
+accounts will be updated too. Each `RewardsReserve` records the reward balance
+to be distributed for a particular block at block height `height` in `Reward Distribution`
+phase. The set of stake rewards paid from `EpochRewardsReserve`
+sysvar accounts form a complete partition of the total stake rewards.
+
+```
+struct EpochRewardReserve {
+   balance: u64,
+   height: u64,
+   hash: Hash
+}
+```
+
+`balance` is total amount of rewards (in lamports) to be distributed in this block.
+`height` is the reward distribution absolute block height.
+`hash` is the hash of all distributed rewards for this block.
+
+
+### Reward Distribution
```

Comment:
An important topic that hasn't really been explicitly discussed is when the lamport supply increases and where those lamports are held until they are distributed. Here's one idea:

1. The bank capitalization increases like it currently does, upon the first bank of the epoch. This is fine because we still credit vote accounts synchronously
2.  The additional lamports that are pending stake account distribution are each held as real lamports in the respective `EpochRewardReserve` accounts. This means that the `EpochRewardReserve::balance` field could be removed entirely, as the account balance itself holds this value. This also requires that each `EpochRewardReserve` account hold a rent-exempt balance, and works around the problem of "where do the lamports held in each EpochRewardReserve account come from"
3. As the each `EpochRewardReserve` account is distributed, the actual sysvar account lamports go to zero and the account is cleaned up normally.